### ISSUE TITLE
idea: relax `applyFixities` type

### ISSUE THREAD

**Issue by int-e** [haskell-suite/haskell-src-exts#318]
Currently `applyFixities` has this type:

    applyFixities :: (Monad m, AppFixity ast) => [Fixity] -> ast SrcSpanInfo -> m (ast SrcSpanInfo)

It would be straightforward to relax it to

    applyFixities :: (Monad m, AppFixity ast, Monoid l) => [Fixity] -> ast l  -> m (ast l)

if one provides a `Monoid` instance for `SrcSpan` and replaces the two calls to `<++>` in the `Language.Haskell.Exts.Fixity` module by `mappend` (or `<>`). This would allow `applyFixities` to be used on ASTs without annotations (using the `()` monoid).

(Technically this only requires a binary operation, so `Semigroup` would be a more appropriate typeclass, but that would add a new dependency.)

**Comment by mpickering** [comment]
Good suggestion thanks.

**Comment by mpickering** [comment]
I investigated this a bit and currently SrcSpan would't admit a lawful
Monoid instance because there is no identity. It seems that the Semigroup
instance would be valid and now the type class is in base maybe we can make
some progress there.

**Comment by pjonsson** [comment]
The Monoid for SrcSpan is #48. I agree with Roman's comments in that ticket.

**Comment by mpickering** [comment]
Explicit representing the empty `SrcSpan` using a sum type means that the fields can't be unpacked. I don't think that it's worth the tradeoff.

**Comment by mpickering** [comment]
It's baffling that the [semigroups](https://hackage.haskell.org/package/semigroups-0.18.2) package has so many dependencies. It should surely be up to the upstream libraries to define the instances? Depending on it would greatly increase the dependency footprint.

**Comment by pjonsson** [comment]
Regarding the performance impact of adding more constructors to SrcSpan--here are the results of `time cabal test` after `cabal clean && cabal configure --enable-tests && cabal build -j4` using ghc 7.8.3. I ran a warmup `cabal test` before doing the timed run. Before:
```
real    3m41.786s
user    3m34.191s
sys     0m5.777s
```
After adding `Empty`:
```
real    3m40.861s
user    3m32.584s
sys     0m5.806s
```
If there were large differences I would expect we'd see them, I doubt the extra 1 second is statistically significant.

As for adding dependencies--we need to keep our footprint small because our users can't just switch to another Haskell parser when they get diamond shaped dependency problems. It doesn't matter if the fault lies in cabal (-install), our sibling packages, or our package--things stil don't build when that happens.

**Comment by mpickering** [comment]
What exact dependency problem are you referring to? This seems like a complete red-herring to me. 

By unpacking the fields allocations will decrease for all users, this is just a consequence of how unpacking works. This seems like a much bigger win to me than adding a `Monoid` instance.

**Comment by pjonsson** [comment]
The point of my measurement numbers were that I can't detect any performance difference from adding a constructor to SrcSpan. In comparison, the strictness annotations on the individual fields made a clear difference.

The dependency problem I'm talking about is when package `A` is trying to use HSE and package `B` at the same time. HSE depends on `something  == version` and package `B` depends on `something == (version - 1)`. Here's one bug report where I had to choose between using either Criterion or Language-C-Quote: https://github.com/mainland/mainland-pretty/pull/4.

**Comment by pjonsson** [comment]
Here's another one of these dependency problems I described: https://github.com/mainland/exception-transformers/issues/3

**Comment by int-e** [comment]
I don't believe that the dependency concerns are justified in the case of `semigroup`, especially when all one does is provide an instance of `Semigroup` and use the `<>` operator. That part of the interface has never changed in all the time of the package's existence. There are transitive dependencies, but `semigroup` only uses basic functionality of those, so again I'm not overly worried about them. And with ghc 8, there is no additional dependency at all.
